### PR TITLE
Seabay/master - finished the project!

### PR DESCRIPTION
**Pull request by kdefliese**
http://seabay.herokuapp.com/

Look at this site, isn't it neat?
Wouldn't you think our test coverage's complete?
Wouldn't you think we're the team
The team who has everything?

Look at this trove, treasures untold
How many merchants can one website hold?
Looking around here you'd think
Sure, Seabay's got everything

We've got gadgets and gizmos aplenty
We've got whozits and whatzits galore
You want thingamabobs? We've got twenty!
But who cares? No big deal,
You want code

I wanna be where the models are
I wanna see, wanna see them querying
Messing around in those
What do you call 'em? Oh, databases

Coding in Ruby you won't get too far
Rails is required for awesome websites
Scrolling along down a
What's that word again?
Responsive page

Up where they CRUD, up where they SCRUM
Up where they code all day in the sun (instead of in the dark)
Wandering free
Wish we could be, done with Betsy

-- with love from Ricky, Sarah, Desiree and Katherine

### CODE REVIEW

File: app/assets/javascripts/carts.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (EliaMG) [inline app/assets/javascripts/carts.coffee:1]: Rails generated files are super cool- but if you have a bunch you don't use, best practice would be to delete the empty files (though soon you will have full javascript files too! :) )

File: spec/models/review_spec.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+RSpec.describe Review, type: :model do
+
+  describe "model validations" do
+    it "requires a value" do
+      empty_item = Review.new
+      expect(empty_item).to be_invalid
+      expect(empty_item.errors.keys).to include :rating
+    end
+  end
+  # does not work - fix!
```
Comment (corinnepingul) [inline spec/models/review_spec.rb:12]: One really cool tool that you can use to make sure you come back to comments like this is by using ruby's quicktips: http://rubyquicktips.com/post/385665023/fixme-todo-and-optimize-code-comments.

Essentially, you write a comment you want to come back to as `# FIXME your comment` all throughout your project (as needed) and then in your project directory in your terminal run `$ rake notes` and it'll show you all these messages and where they are located in your project. (more detail in the link above)

File: spec/models/review_spec.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+require 'rails_helper'
+
+RSpec.describe Review, type: :model do
+
+  describe "model validations" do
+    it "requires a value" do
```
Comment (corinnepingul) [inline spec/models/review_spec.rb:6]: I would replace "value" here with "rating" to show exactly which attribute of the review model you are validating.

File: app/models/product.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  validates :price, presence: true, numericality: { greater_than: 0 }
+
+  has_attached_file :image, styles: { large: "600x600>", medium: "300x300>", thumb: "100x100>" }, :default_url => "/images/:style/missing.png"
+  validates_attachment_content_type :image, content_type: /\Aimage\/.*\Z/
+
+
+
+  def average_review
+    reviews.average(:rating) || 0
+    # if @reviews.length != 0
```
Comment (sallyamoore) [inline app/models/product.rb:19]: It's a good idea to either delete deprecated code or add a comment saying why you're keeping it. :grinning:

File: app/models/product.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+class Product < ActiveRecord::Base
+  belongs_to :merchant
+  has_many :reviews
+  has_many :order_items
+  has_many :orders, through: :order_items
+  has_and_belongs_to_many :categories
+
+  validates_associated :reviews
```
Comment (sallyamoore) [inline app/models/product.rb:8]: Nice validations!!!

File: app/controllers/products_controller.rb [hunk 6]
```diff
@@ -0,0 +1,103 @@
+class ProductsController < ApplicationController
+  before_action :require_login, only: [:index, :edit, :new]
+  before_action :find_product, only: [:show, :edit, :retire, :activate]
+
+  def index
```
Comment (sallyamoore) [inline app/controllers/products_controller.rb:5]: When you have methods like this, you can put them on one line with `def index; end`!

File: app/controllers/products_controller.rb [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+  def activate
+    @product.status = "Active"
+    @product.save!
+    redirect_to merchant_products_path(@product.merchant_id)
+  end
+
+  private
+
+  def current_merchant_product?
```
Comment (sallyamoore) [inline app/controllers/products_controller.rb:78]: Comment on the purpose of methods like this one would be helpful to others looking at your code. :heart: comments! Always Be Commenting! :smile:

File: app/controllers/products_controller.rb [hunk 8]
```diff
@@ -0,0 +1,103 @@
+class ProductsController < ApplicationController
+  before_action :require_login, only: [:index, :edit, :new]
+  before_action :find_product, only: [:show, :edit, :retire, :activate]
```
Comment (sallyamoore) [inline app/controllers/products_controller.rb:3]: Excellent use of before actions!!!

File: app/views/products/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+      <h3>Reviews:</h3>
+      <table class="table table-bordered">
+        <tbody>
+          <tr>
+            <th>Rating</th>
+            <th>Comment</th>
+          </tr>
+          <% @reviews.each do |review| %>
+              <% each_review = review.rating %>
+              <tr><td><% each_review.times do |rating| %>
```
Comment (corinnepingul) [inline app/views/products/show.html.erb:59]: For this iteration that prints the number of starfish for the review value, you don't need the pipes `|rating|` since each iteration will be printing out the same thing (the starfish image). 

Scroll down on this link for an example of using the `.times` method with vs. without the iteration variable: http://www.dotnetperls.com/iterator-ruby

File: app/controllers/orders_controller.rb [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+  def checkout
+    # get @current_order info from carts controller
+    @cart_items = session[:cart]
+    # @cart_items looks like { "1" => 2 }
+    # want instance_vars @p1, @p2, etc. set equal to the keys of the hash
+    @order = Order.new
+    @order_item = OrderItem.new
+    @products = []
+    if !@cart_items.nil?
```
Comment (EliaMG) [inline app/controllers/orders_controller.rb:15]: you can do `if @cart_items` here, try to avoid double negatives as much as possible as it is harder to understand the true/false case. Same applies for line 50 :smile:

File: app/views/products/show.html.erb [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+                <%= image_tag("starfish.jpg", size: "15x15", alt: "Starfish rating") %>
+              <% end %></td>
+              <td><%= review.description %></td>
+            </tr>
+          <% end %>
+          <% if @reviews.count == 0 %>
```
Comment (corinnepingul) [inline app/views/products/show.html.erb:65]: I would recommend to do this type of `if` check first with the `@reviews.each` iteration that's in the above lines in an `else` block so that none of that code gets run if you don't have any reviews to iterate over.

File: app/views/products/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    <div class="col-md-12">
+      <h3>Description:</h3>
+      <p><%= @product.description %></p>
+      <% if @product.description.length == 0 %>
+      <p>For just <%= number_to_currency(@product.price) %>, this ocean-related item can soon be yours!</p>
+      <% end %>
+      <br>
+  <h3>Average rating:</h3>
+  <% if @reviews.length != 0 %>
+    <% @average.times do |average| %>
```
Comment (corinnepingul) [inline app/views/products/show.html.erb:39]: See comment below about using the `.times` method without the iteration variable (what's inside the pipes)

File: app/views/products/all_products.html.erb [hunk 13]
```diff
@@ -0,0 +1,25 @@
+<div class="container">
+  <%= render 'layouts/navbar' %>
+<h2>All Products</h2>
+<% @products.each_with_index do |product, index| %>
+  <% if product.status == "Active" && product.quantity != 0 %>
+    <% if index % 6 == 0 %>
```
Comment (sallyamoore) [inline app/views/products/all_products.html.erb:6]: I think this section of the code is intended to put 6 product cards in each visual row. It's fantastic that you're trying to control the appearance of each visual row on the page! Unfortunately, inspecting the page reveals that each row winds up getting nested inside the row(s) that came before it. Also, using `index % 6` to split these probably isn't the best way. I'd probably use bootstrap cols without the rows. Each card would also have to be uniform size (width and height) for this to work.

File: app/controllers/orders_controller.rb (hunk 10 above)
Comment (EliaMG) [inline app/controllers/orders_controller.rb:15]: If you didn't want to set the cart to nil if it is empty (carts_controller#index:line18), you could also do an `unless @cart_items.empty?` check here-- will try to comment more about that in the carts_controller.

File: app/controllers/orders_controller.rb [hunk 14]
```diff
@@ -0,0 +1,182 @@
+class OrdersController < ApplicationController
+  before_action :require_login, only: [:index, :pending, :cancelled, :paid, :completed, :show]
+  before_action :get_order, only: [:confirm, :show]
+  before_action :get_order_items, only: [:index, :status]
+  before_action :get_order_item_revenue, only: :show
+
+  def checkout
+    # get @current_order info from carts controller
+    @cart_items = session[:cart]
+    # @cart_items looks like { "1" => 2 }
```
Comment (EliaMG) [inline app/controllers/orders_controller.rb:10]: thanks for this comment! It really helped me understand why you called k.to_i on Line 17 :heart:

File: app/views/shared/_productdetail.html.erb [hunk 15]
```diff
@@ -0,0 +1,31 @@
+  <%= form_for [@current_merchant, @product], html: { multipart: true }, url: { action: @action}, method: @method, :html => { :class => "form"} do |f| %>
+    <div class="form-group">
+      <%= f.label :name %>
+      <%= f.text_field :name, :required => true, class:"form-control" %>
+    </div>
+    <div class="form-group">
+      <%= f.label :price %>
+      <div class="input-group">
+       <div class="input-group-addon">$</div>
+       <%= f.number_field :price, in: 0.01..1000000000, step: 0.01, :required => true, class:"form-control" %>
```
Comment (sallyamoore) [inline app/views/shared/_productdetail.html.erb:10]: You could also use `value: number_to_currency()` here.

**Comment by corinnepingul** [comment]
Hi friends! 

@EliaMG, @sallyamoore, and I have started the code review for your betsy project! 

We wanted to let you know that we're about a third to halfway through and will finish by this coming Tuesday evening. This is just a heads up for you in case you want to wait to have the complete review before starting to look at our comments. If you do feel like looking at our comments before the end of Tuesday, feel free to comment/ask any questions if you have them!

Hope all is well with class :)

File: app/views/products/show.html.erb [hunk 16]
```diff
@@ -0,0 +1,82 @@
+<div class="container">
+  <%= render 'layouts/navbar' %>
+
+  <% if !flash[:error].nil? %>
```
Comment (sallyamoore) [inline app/views/products/show.html.erb:4]: Tends to be easier to read/understand `unless thing` rather than `if !thing`, especially when combined with `.nil?` or `.empty?`

File: app/controllers/order_items_controller.rb [hunk 17]
```diff
@@ -0,0 +1,22 @@
+class OrderItemsController < ApplicationController
+  before_action :require_login
+  before_action :find_orderitem, only: [:edit, :update]
```
Comment (EliaMG) [inline app/controllers/order_items_controller.rb:3]: Good use of the before_action here! I especially like this in a short controller as it doesn't get lost-- I have had issues with before_actions (dependent ones in another controller) in my internship that made it really hard to follow/debug-- I thought this article was an interesting take: http://blog.thefrontiergroup.com.au/2014/02/before_action_an_anti_pattern/ (good to read with a grain of salt and generally follow the style/convention of Ada class/your internship/job)

File: app/views/products/show.html.erb (hunk 12 above)
Comment (sallyamoore) [inline app/views/products/show.html.erb]: you can remove |average| here

File: app/controllers/orders_controller.rb [hunk 18]
```diff
@@ -0,0 +1,182 @@
+class OrdersController < ApplicationController
+  before_action :require_login, only: [:index, :pending, :cancelled, :paid, :completed, :show]
```
Comment (EliaMG) [inline app/controllers/orders_controller.rb:2]: You may want to use an :except here since require_login is used on most of the actions.

File: spec/controllers/products_controller_spec.rb [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+  describe "GET #index" do
+    it "is successful" do
+      merchant.authenticate(session_data)
+      session[:user_id] = merchant.id
+      get :index, merchant_id: @product.merchant_id
+      expect(response.status).to eq 200
+    end
+  end
+
+  describe "GET 'new'" do
```
Comment (sallyamoore) [inline spec/controllers/products_controller_spec.rb:80]: Consistency in how things are named/referenced is helpful for those consuming your code. I like your use of the pattern `describe "VERB #method"` in your specs and suggest using that here and on line 91 as well.

File: spec/controllers/products_controller_spec.rb [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+      get :new, merchant_id: @product.merchant_id
+      expect(response.status).to eq 200
+      expect(subject).to render_template :new
+    end
+  end
+
+  describe "#show" do
+    before :each do
+      Review.create(good_review[:review])
+      Review.create(rating: 1, description: "Aight", product_id: @product.id)
```
Comment (sallyamoore) [inline spec/controllers/products_controller_spec.rb:94]: This is the sort of thing :sparkles:FactoryGirl:sparkles: will do for you soon!!!

File: spec/controllers/products_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+  { merchant_id: 1,
+    product: {
+      name: "Dogfish",
+      price: 2,
+      description: "new description"
+    }
+  }
+  end
+
+  let(:update_params) do
```
Comment (sallyamoore) [inline spec/controllers/products_controller_spec.rb:18]: An alternative to making 3 sets of params here would be to make one valid set and change an attribute in the test to make it different or invalid. You'll follow that pattern when you're using :sparkles:FactoryGirl:sparkles:!

File: spec/controllers/products_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+    expect(subject).to render_template :edit
+    end
+  end
+
+  describe "PATCH #update" do
+    it "goes to the product index page when successful" do
+      patch :update, update_params
+      expect(subject).to redirect_to merchant_products_path(@product.merchant_id)
+    end
+    it "renders the edit view when unsuccessful" do
```
Comment (sallyamoore) [inline spec/controllers/products_controller_spec.rb:137]: Putting a blank line between blocks helps others read your code!

File: spec/controllers/products_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+      patch :update, bad_params
+      expect(subject).to render_template :edit,(@product.id)
+    end
+  end
+
+  describe "DESTROY #destroy" do
```
Comment (sallyamoore) [inline spec/controllers/products_controller_spec.rb:143]: The HTML verb here is `DELETE` not `DESTROY` :smile:

File: spec/controllers/products_controller_spec.rb [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+    end
+
+    it "can't be created if merchant tries to review their own product" do
+      # log in merchant
+      merchant.authenticate(session_data)
+      session[:user_id] = merchant.id
+
+      post :create_review, good_review
+      expect(Review.count).to eq(0)
+      expect(subject).to redirect_to product_path(1)
```
Comment (sallyamoore) [inline spec/controllers/products_controller_spec.rb:164]: Should attempting this raise an error? If so, that would be good to test.

Comment (sallyamoore) [inline spec/controllers/products_controller_spec.rb:164]: Another option is to check if your post changes Review.count. That would look something like:

``` ruby
expect { 
  post :create_review, good_review 
}.to change(Review, :count).by(1)
```

That way your test is more targeted... What you really care about is that the count changes, not that it is an absolute value.

File: spec/controllers/products_controller_spec.rb [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+    end
+  end
+
+  describe "retire a product" do
+    it "redirects to merchant index page" do
+      merchant.authenticate(session_data)
+      session[:user_id] = merchant.id
+      patch :retire, update_params
+      expect(subject).to redirect_to merchant_products_path(@product.merchant_id)
```
Comment (sallyamoore) [inline spec/controllers/products_controller_spec.rb:186]: Since this controller method also sets the product's status to "Retired", it would be good to test that behavior as well.

File: spec/controllers/products_controller_spec.rb [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+  end
+
+  describe "activate a product" do
+    it "redirects to merchant index page" do
+      merchant.authenticate(session_data)
+      session[:user_id] = merchant.id
+      patch :activate, update_params
+      expect(subject).to redirect_to merchant_products_path(@product.merchant_id)
+    end
+  end
```
Comment (sallyamoore) [inline spec/controllers/products_controller_spec.rb:197]: Since this controller method also sets the product's status to "Active", it would be good to test that behavior as well.

File: app/controllers/categories_controller.rb [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+    @category = Category.find(params[:id])
+  end
+
+  private
+
+  def category_params
+    params.permit(category:[:id, :name, :image, :photo_url])
+  end
+
+end
```
Comment (sallyamoore) [inline app/controllers/categories_controller.rb:29]: Very readable, straightforward controller! Nice work! :fireworks:

File: spec/controllers/products_controller_spec.rb [hunk 28]
```diff
@@ -0,0 +1,198 @@
...
+
+  let(:bad_params) do
+    { merchant_id: 1,
+      id: 1,
+      product: {
+        name: "",
+      }
+    }
+  end
+
```
Comment (sallyamoore) [inline spec/controllers/products_controller_spec.rb:37]: I'd suggest using `valid_<object_name>` and `invalid_<object_name>` rather than `good_<object_name>` and `bad_<object_name>`. Descriptive naming is _super_ helpful when you're working in a larger code base... or looking at code you wrote a while ago. :wink: 

Also see my almost identical comment in the product model specs.

File: spec/models/product_spec.rb [hunk 29]
```diff
@@ -0,0 +1,45 @@
+require 'rails_helper'
+
+RSpec.describe Product, type: :model do
+
+  empty_product = Product.new
+
+  let(:good_product) do
```
Comment (sallyamoore) [inline spec/models/product_spec.rb:7]: I'd suggest using `valid_product` and `invalid_product` rather than `good_product` and `bad_product`. Descriptive naming is _super_ helpful when you're working in a larger code base... or looking at code you wrote a while ago. :wink:

File: spec/models/product_spec.rb [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+      expect(bad_price).to be_invalid
+      expect(bad_price.errors.keys).to include (:price)
+    end
+    it "expects price to be greater than 0" do
+      bad_price = Product.create({ name: "Dolphin", price: 0, description: "description"})
+      expect(bad_price).to be_invalid
+      expect(bad_price.errors.keys).to include (:price)
+    end
+  end
+end
```
Comment (sallyamoore) [inline spec/models/product_spec.rb:45]: Nicely tested model validations!!! :star:

File: app/views/categories/show.html.erb [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+<% else %>
+  50% 20%;'>
+<% end %>
+  <h2><%= @category.name.upcase %></h2>
+</div>
+<div class="container container-home-page">
+  <%= render 'layouts/navbar' %>
+  <% @category.products.each_with_index do |product, i| %>
+  <% if product.status == "Active" && product.quantity != 0 %>
+      <% if i + 1 % 3 == 0 %>
```
Comment (sallyamoore) [inline app/views/categories/show.html.erb:22]: See my note on line 6 in views/products/all_products.html.erb about using % to control visual display. It can lead to unexpected behavior (e.g., b/c not all the visual cards are the same size), so I'd use bootstrap cols instead.

File: app/views/categories/show.html.erb [hunk 32]
```diff
@@ -0,0 +1,45 @@
...
+        <div class="row">
+      <% end %>
+      <div class="col-sm-4 col-md-4">
+        <div class="thumbnail">
+          <% if product.image_file_name.nil? %>
+            <%= image_tag(product.photo_url, :size => "300") %>
+          <% else %>
+          <%= image_tag(product.image.url(:medium))%>
```
Comment (sallyamoore) [inline app/views/categories/show.html.erb:30]: indent in a level on line 30 :)

File: app/views/categories/show.html.erb [hunk 33]
```diff
@@ -0,0 +1,45 @@
...
+          <% end %>
+          <div class="caption">
+            <%= link_to product.name, product_path(product.id), class: "h3" %>
```
Comment (sallyamoore) [inline app/views/categories/show.html.erb:33]: Could be confusing to use an HTML tag as your class name. Instead, perhaps something more descriptive (e.g., "link" if you use this class to style all links)?

File: spec/controllers/categories_controller_spec.rb [hunk 34]
```diff
@@ -0,0 +1,64 @@
+require 'rails_helper'
+
+RSpec.describe CategoriesController, type: :controller do
+  let(:good_params) do
```
Comment (sallyamoore) [inline spec/controllers/categories_controller_spec.rb:4]: see note on line 37 in spec/controllers/products_controller_spec.rb re: naming

File: spec/controllers/categories_controller_spec.rb [hunk 35]
```diff
@@ -0,0 +1,64 @@
...
+      expect(subject).to render_template :new
+    end
+  end
+
+  describe "POST 'create'" do
+    it "successful create redirects to category show page" do
+      expect(Category.all.length).to eq 0
+      post :create, good_params
+      expect(response.status).to eq 302
+      expect(Category.all.length).to eq 1
```
Comment (sallyamoore) [inline spec/controllers/categories_controller_spec.rb:43]: You could use `expect{ action }.to change(Category.count)` here! See second note on line 164 of spec/controllers/products_controller_spec.rb.

File: app/controllers/orders_controller.rb [hunk 36]
```diff
@@ -0,0 +1,182 @@
...
+      session[:cart] = nil
+      redirect_to order_confirm_path(@order.id)
+    else
+      # tells guest which products were sold out and removed from cart
+      if !@inventory_errors.empty?
+        flash[:error] = @inventory_errors.join(", ")
+      end
+      redirect_to orders_checkout_path
+    end
+  end
```
Comment (EliaMG) [inline app/controllers/orders_controller.rb:75]: This is a very long and complex method. Can you think of ways to pull out chunks of code into model method(s)? It is a lot of logic to track within a controller action.

File: app/controllers/orders_controller.rb [hunk 37]
```diff
@@ -0,0 +1,182 @@
...
+    # get all order items for the order that belong to signed in merchant
+    @order_items = []
+
+    @all_order_items.each do |oi|
+      if oi.product.merchant_id == current_merchant.id
+        @order_items.push(oi)
+      end
+    end
+
+    # find total revenue for the signed in merchant's order items
```
Comment (EliaMG) [inline app/controllers/orders_controller.rb:178]: nice comments in your private methods

File: app/models/order.rb [hunk 38]
```diff
@@ -0,0 +1,21 @@
+class Order < ActiveRecord::Base
+  has_many :order_items
+  has_many :products, through: :order_items
+
+  validates_presence_of :name, :bill_zip, :cc_exp, :cc_num, :city, :email, :sec_code, :state, :street, :zip
+  validates :city, length: { maximum: 30 }
+  # validates :cc_num, length: { is: 4 }
+  validates :status, inclusion: { in: %w(Pending Paid Complete Cancelled) }
```
Comment (EliaMG) [inline app/models/order.rb:8]: I'm not sure city length is the most important validation, and it would be a good idea to validate credit card number length, zip code length/are numbers, cc expiration date etc.   It was pretty easy to break the app just by typing a long string of numbers into the zip field (it breaks postgres max int size if not a bignum). 
I know validations can be a late priority, but it is good to think about how to best do them at the model level where possible.

File: config/routes.rb [hunk 39]
```diff
@@ -1,6 +1,42 @@
...
+  post 'orders/checkout' => 'orders#create'
+  get 'orders/:id/confirm' => 'orders#confirm', as: :order_confirm
+  get "merchants/:id/orders/show/:status" => 'orders#status', as: :status_orders
+
+  # merchants routes
+  get "merchants/home" => "merchants#home", as: :merchant_home
+
+  resources :merchants, :except => :destroy do
+    resources :products, :except => :show
+    resources :orders, :except => [:destroy, :new, :create]
```
Comment (EliaMG) [inline config/routes.rb:37]: Nice job removing unused routes :smiling_imp:

File: db/migrate/20151208193058_create_orders.rb [hunk 40]
```diff
@@ -0,0 +1,20 @@
...
+      t.string :status
+      t.string :name
+      t.string :email
+      t.string :street
+      t.string :city
+      t.string :state
+      t.integer :zip
+      t.integer :cc_num
+      t.date :cc_exp
+      t.integer :sec_code
```
Comment (EliaMG) [inline db/migrate/20151208193058_create_orders.rb:14]: Though this is a mock app, I will mention that you would never want to persist all this information in one table IRL. It would make it much too easy for a person's credit card information to be stolen and used.  If you wanted to store something to know which credit card, you could just store the last 4 of what was passed in via the form. This is why there is such big business for paypal, square etc so that small sites never have to worry about persisting, storing, or losing customer PII! :smile:

File: app/controllers/merchants_controller.rb [hunk 41]
```diff
@@ -0,0 +1,32 @@
+class MerchantsController < ApplicationController
+  before_action :require_login, only: [:home]
```
Comment (Brennaleker) [inline app/controllers/merchants_controller.rb:2]: nice use of a before action! :star:

File: app/controllers/merchants_controller.rb [hunk 42]
```diff
@@ -0,0 +1,32 @@
...
+      end
+  end
+
+  def home
+  end
+
+  def show
+    @merchant = Merchant.find(params[:id])
+    @products = @merchant.products
+    @retiredproducts = @products.where(status:"Retired")
```
Comment (Brennaleker) [inline app/controllers/merchants_controller.rb:24]: nice clean query

File: app/models/merchant.rb [hunk 43]
```diff
@@ -0,0 +1,8 @@
+class Merchant < ActiveRecord::Base
+  has_many :products
+  has_secure_password
+  validates :username, presence: true, uniqueness: true
+  validates :email, presence: true, uniqueness: true
```
Comment (Brennaleker) [inline app/models/merchant.rb:5]: nice validation grouping, you could validate the formatting of email using [regex](http://stackoverflow.com/questions/4776907/what-is-the-best-easy-way-to-validate-an-email-address-in-ruby)

File: app/views/merchants/show.html.erb [hunk 44]
```diff
@@ -0,0 +1,41 @@
+<div class="container">
+  <%= render 'layouts/navbar' %>
+  <div class="row">
+    <div class="col-md-12">
+      <h1>Browse <%= @merchant.username %>'s Products</h1>
+    </div>
+  </div>
+
+  <div class="row">
+    <% if @products.count == 0 || @products.count == @retiredproducts.count %>
```
Comment (Brennaleker) [inline app/views/merchants/show.html.erb:10]: Great info message

File: app/views/merchants/new.html.erb [hunk 45]
```diff
@@ -0,0 +1,37 @@
+<div class="container fill-register">
+  <%= render 'layouts/navbar' %>
+  <% if !@merchant.errors.empty? %>
```
Comment (Brennaleker) [inline app/views/merchants/new.html.erb:3]: this is a great way to inform your user of any errors, another option is [flash alerts](http://api.rubyonrails.org/classes/ActionDispatch/Flash.html)

File: app/views/merchants/show.html.erb [hunk 46]
```diff
@@ -0,0 +1,41 @@
...
+        <% if product.quantity != 0 %>
+          <% if product.status == "Active" %>
+            <% if i + 1 % 3 == 0 %>
+              <div class="row">
+            <% end %>
+            <div class="col-sm-4 col-md-4">
+              <div class="thumbnail">
+                <% if product.image_file_name.nil? %>
+                  <%= image_tag(product.photo_url, :size => "300x169") %>
+                <% else %>
```
Comment (Brennaleker) [inline app/views/merchants/show.html.erb]: indent in a level on line 25

File: app/views/merchants/show.html.erb [hunk 47]
```diff
@@ -0,0 +1,41 @@
...
+                <%= image_tag(product.image.url(:medium))%>
```
Comment (Brennaleker) [inline app/views/merchants/show.html.erb:26]: indent in a level on 26

File: spec/models/merchant_spec.rb [hunk 48]
```diff
@@ -0,0 +1,43 @@
+require 'rails_helper'
+
+RSpec.describe Merchant, type: :model do
+  let(:empty_merchant) do
```
Comment (Brennaleker) [inline spec/models/merchant_spec.rb:4]: very smart to create one empty merchant to use in all tests for presence validations :+1:

**Comment by acmei** [comment]
OMG I just read the comment for this PR. :clap: :sparkling_heart: :musical_note:
